src/types/params/plain_secret.rs

```diff
@@ -139,18 +139,63 @@ impl<'a> PlainSecretParamsRef<'a> {
                     ref hash,
                     ref alg_sym,
                     ..
-                } => match *curve {
+                } => match curve {
                     ECCCurve::Curve25519 => {
-                        ensure!(d.len() <= 32, "invalid secret");
+                        const SEC_LEN: usize = ECCCurve::Curve25519.secret_key_length();
 
-                        let mut secret = [0u8; 32];
-                        secret[32 - d.len()..].copy_from_slice(d.as_bytes());
+                        ensure!(d.len() <= SEC_LEN, "invalid secret");
 
-                        Ok(SecretKeyRepr::ECDH(crate::crypto::ecdh::SecretKey {
-                            oid: curve.oid(),
+                        let mut secret = [0u8; SEC_LEN];
+                        secret[SEC_LEN - d.len()..].copy_from_slice(d.as_bytes());
+
+                        Ok(SecretKeyRepr::ECDH(
+                            crate::crypto::ecdh::SecretKey::Curve25519Legacy {
+                                secret,
+                                hash: *hash,
+                                alg_sym: *alg_sym,
+                            },
+                        ))
+                    }
+                    ECCCurve::P256 => {
+                        const SEC_LEN: usize = ECCCurve::P256.secret_key_length();
+
+                        ensure!(d.len() <= SEC_LEN, "invalid secret");
+
+                        let mut secret = [0u8; SEC_LEN];
+                        secret[SEC_LEN - d.len()..].copy_from_slice(d.as_bytes());
```

Comment:
maybe add a helper, sth like this

```rust
fn pad_key<const SIZE: usize>(val: &[u8]) -> Result<[u8; SIZE]> {
  ensure!(d.len() <= SIZE, "invalid key size");
  let mut key = [0u8; SIZE];
  key[SIZE - val.len()..].copy_from_slice(val);
  Ok(key)
}
```

Reply:
Right.

(Aside: I was surprised to notice that the compiler accepts calls to this kind of function without an explicit value for `SIZE`. It's legal to call just `pad_key(&val)`.
I failed to find Rust documentation that explains this, and if there is a way to tell the compiler that the parameter must be explicitly set.)